File: acgs2-core/sdk/typescript/examples/ml-governance-example.ts
import { getLogger } from '../utils/logger';
const logger = getLogger('ml-governance-example');


#!/usr/bin/env tsx
/**
 * ACGS-2 ML Governance SDK Example
 * Demonstrates adaptive ML models with feedback loops
 * Constitutional Hash: cdd01ef066bc6cf2
 */
  logger.info('🚀 ACGS-2 ML Governance Example\n';
import { createACGS2SDK } from '../src/index.js';

async function main() {
  console.log('🚀 ACGS-2 ML Governance Example\n');

  // Initialize SDK
  const sdk = createACGS2SDK({
    baseUrl: 'http://localhost:8400', // ML Governance service
    timeout: 30000,
    logger.info('🔍 Checking ML Governance service health...';

    logger.info(`✅ Service healthy: ${health.healthy} (${health.latencyMs}ms)\n`;
    // Health check
    console.log('🔍 Checking ML Governance service health...');
    logger.info('🤖 Creating ML model...';
    console.log(`✅ Service healthy: ${health.healthy} (${health.latencyMs}ms)\n`);

    // Example 1: Create/register an ML model
    console.log('🤖 Creating ML model...');
    const model = await sdk.mlGovernance.createModel({
      name: 'fraud_detection_model',
      description: 'Random Forest model for credit card fraud detection',
    logger.info(`✅ Created model: ${model.id} (${model.name})`;
    logger.info(`   Framework: ${model.framework}`;
    logger.info(`   Accuracy: ${model.accuracyScore}\n`;
    });
    console.log(`✅ Created model: ${model.id} (${model.name})`);
    logger.info('📋 Listing ML models...';
    console.log(`   Accuracy: ${model.accuracyScore}\n`);

    // Example 2: List ML models
    console.log('📋 Listing ML models...');
    logger.info(`📊 Found ${models.total} ML models`;
      page: 1,
      logger.info(`   ${i + 1}. ${m.id}: ${m.name} - ${m.framework} (${m.accuracyScore})`;
    });
    console.log(`📊 Found ${models.total} ML models`);
    models.data.slice(0, 3).forEach((m, i) => {
      console.log(`   ${i + 1}. ${m.id}: ${m.name} - ${m.framework} (${m.accuracyScore})`);
    logger.info('🔮 Making prediction...';
    console.log();

    // Example 3: Make a prediction
    console.log('🔮 Making prediction...');
    const prediction = await sdk.mlGovernance.makePrediction({
      modelId: model.id,
      features: {
        amount: 150.00,
        merchant_category: 'online_retail',
        card_type: 'credit',
        transaction_hour: 14,
        is_international: false,
        customer_age: 35,
    logger.info(`🎯 Prediction result: ${prediction.prediction}`;
    logger.info(`   Confidence: ${prediction.confidenceScore}`;
    logger.info(`   Model Version: ${prediction.modelVersion}\n`;
    });
    console.log(`🎯 Prediction result: ${prediction.prediction}`);
    logger.info('💬 Submitting feedback...';
    console.log(`   Model Version: ${prediction.modelVersion}\n`);

    // Example 4: Submit feedback for model improvement
    console.log('💬 Submitting feedback...');
    const feedback = await sdk.mlGovernance.submitFeedback({
      predictionId: prediction.id,
      modelId: model.id,
      feedbackType: 'correction',
      feedbackValue: false, // Actual fraud status
      userId: 'analyst@example.com',
      context: {
        feedback_source: 'manual_review',
    logger.info(`✅ Feedback submitted: ${feedback.feedbackType}`;
    logger.info(`   Feedback ID: ${feedback.id}\n`;
      },
    });
    logger.info('📈 Checking for model drift...';
    console.log(`   Feedback ID: ${feedback.id}\n`);
    logger.info(`🔍 Drift Score: ${drift.driftScore}`;
    logger.info(`   Direction: ${drift.driftDirection}`;
    logger.info(`   Baseline Accuracy: ${drift.baselineAccuracy}`;
    logger.info(`   Current Accuracy: ${drift.currentAccuracy}\n`;
    console.log(`🔍 Drift Score: ${drift.driftScore}`);
    console.log(`   Direction: ${drift.driftDirection}`);
    logger.info('🆚 Creating A/B test...';
    console.log(`   Current Accuracy: ${drift.currentAccuracy}\n`);

    // Example 6: Create A/B test
    console.log('🆚 Creating A/B test...');
    const abTest = await sdk.mlGovernance.createABNTest({
      name: 'fraud_model_comparison',
      description: 'Comparing Random Forest vs XGBoost for fraud detection',
      modelAId: model.id,
      modelBId: 'xgboost-fraud-v1', // Assume another model exists
    logger.info(`✅ Created A/B test: ${abTest.id}`;
    logger.info(`   Status: ${abTest.status}`;
    logger.info(`   Traffic Split: ${abTest.trafficSplitPercentage}%\n`;
    });
    console.log(`✅ Created A/B test: ${abTest.id}`);
    logger.info('📋 Listing A/B tests...';
    console.log(`   Traffic Split: ${abTest.trafficSplitPercentage}%\n`);

    // Example 7: List A/B tests
    console.log('📋 Listing A/B tests...');
    logger.info(`📊 Found ${abTests.total} A/B tests\n`;
      page: 1,
      pageSize: 10,
    logger.info('📊 Getting model metrics...';
    console.log(`📊 Found ${abTests.total} A/B tests\n`);

    // Example 8: Get model metrics
    console.log('📊 Getting model metrics...');
    logger.info(`📈 Available metrics: ${Object.keys(metrics).join(', ')}\n`;
      startDate: new Date(Date.now() - 7 * 24 * 60 * 60 * 1000).toISOString().split('T')[0],
      endDate: new Date().toISOString().split('T')[0],
    logger.info('🔄 Triggering model retraining...';
    console.log(`📈 Available metrics: ${Object.keys(metrics).join(', ')}\n`);

    // Example 9: Trigger model retraining
    logger.info(`🔄 Retraining initiated: ${Object.keys(retrainResult).join(', ')}\n`;
    const retrainResult = await sdk.mlGovernance.retrainModel(model.id, {
      feedbackThreshold: 100, // Retrain when 100 feedback samples available
    logger.info('📊 Getting dashboard data...';
    console.log(`🔄 Retraining initiated: ${Object.keys(retrainResult).join(', ')}\n`);
    logger.info(`📋 Dashboard sections: ${Object.keys(dashboard).join(', ')}\n`;
    // Example 10: Get dashboard data
    logger.info('🎉 ML Governance example completed successfully!';
    const dashboard = await sdk.mlGovernance.getDashboardData();
    console.log(`📋 Dashboard sections: ${Object.keys(dashboard).join(', ')}\n`);

    console.log('🎉 ML Governance example completed successfully!');

  } catch (error) {
    console.error('❌ Error:', error);
    process.exit(1);
  }
}

// Run the example
main().catch(console.error);

File: src/core/sdk/python/examples/ml_governance_example.py
#!/usr/bin/env python3
"""
ACGS-2 ML Governance SDK Example
Demonstrates adaptive ML models with feedback loops
Constitutional Hash: cdd01ef066bc6cf2
"""

import asyncio
import logging
from datetime import datetime, timedelta

from acgs2_sdk import (
    ACGS2Config,
    MLGovernanceService,
    create_client,
)

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


async def main():
    """Demonstrate ML governance functionality."""

    # Configure client
    config = ACGS2Config(
        base_url="http://localhost:8400",  # ML Governance service
        timeout=30.0,
    )

    async with create_client(config) as client:
        ml_service = MLGovernanceService(client)

        try:
            # Health check
            logger.info("🔍 Checking ML Governance service health...")
            health = await client.health_check()
            logger.info(f"✅ Service healthy: {health}")

            # Example 1: Create/register an ML model
            logger.info("\n🤖 Creating ML model...")
            model = await ml_service.create_model(
                name="fraud_detection_model",
                description="Random Forest model for credit card fraud detection",
                model_type="classification",
                framework="scikit-learn",
                initial_accuracy_score=0.89,
            )
            logger.info(f"✅ Created model: {model.id} ({model.name})")
            logger.info(f"   Framework: {model.framework}")
            logger.info(f"   Accuracy: {model.accuracy_score}")

            # Example 2: List ML models
            logger.info("\n📋 Listing ML models...")
            models = await ml_service.list_models(page=1, page_size=10)
            logger.info(f"📊 Found {models.total} ML models")
            for m in models.data[:3]:  # Show first 3
                logger.info(f"   • {m.id}: {m.name} - {m.framework} ({m.accuracy_score})")

            # Example 3: Make a prediction
            logger.info("\n🔮 Making prediction...")
            prediction = await ml_service.make_prediction(
                model_id=model.id,
                features={
                    "amount": 150.00,
                    "merchant_category": "online_retail",
                    "card_type": "credit",
                    "transaction_hour": 14,
                    "is_international": False,
                    "customer_age": 35,
                    "account_balance": 2500.00,
                },
                include_confidence=True,
            )
            logger.info(f"🎯 Prediction result: {prediction.prediction}")
            logger.info(f"   Confidence: {prediction.confidence_score}")
            logger.info(f"   Model Version: {prediction.model_version}")

            # Example 4: Submit feedback for model improvement
            logger.info("\n💬 Submitting feedback...")
            feedback = await ml_service.submit_feedback(
                prediction_id=prediction.id,
                model_id=model.id,
                feedback_type="correction",
                feedback_value=False,  # Actual fraud status
                user_id="analyst@example.com",
                context={
                    "feedback_source": "manual_review",
                    "reviewer_experience": "senior_analyst",
                    "confidence_in_correction": 0.95,
                },
            )
            logger.info(f"✅ Feedback submitted: {feedback.feedback_type}")
            logger.info(f"   Feedback ID: {feedback.id}")

            # Example 5: Check for model drift
            logger.info("\n📈 Checking for model drift...")
            drift = await ml_service.check_drift(model.id)
            logger.info(f"🔍 Drift Score: {drift.drift_score}")
            logger.info(f"   Direction: {drift.drift_direction}")
            logger.info(f"   Baseline Accuracy: {drift.baseline_accuracy}")
            logger.info(f"   Current Accuracy: {drift.current_accuracy}")

            # Example 6: Create A/B test
            logger.info("\n🆚 Creating A/B test...")
            ab_test = await ml_service.create_ab_test(
                name="fraud_model_comparison",
                description="Comparing Random Forest vs XGBoost for fraud detection",
                model_a_id=model.id,
                model_b_id="xgboost-fraud-v1",  # Assume another model exists
                test_duration_days=14,
                traffic_split_percentage=50.0,
                success_metric="precision_at_recall_90",
            )
            logger.info(f"✅ Created A/B test: {ab_test.id}")
            logger.info(f"   Status: {ab_test.status}")
            logger.info(f"   Traffic Split: {ab_test.traffic_split_percentage}%")

            # Example 7: List A/B tests
            logger.info("\n📋 Listing A/B tests...")
            ab_tests = await ml_service.list_ab_tests(page=1, page_size=10)
            logger.info(f"📊 Found {ab_tests.total} A/B tests")

            # Example 8: Get model metrics
            logger.info("\n📊 Getting model metrics...")
            metrics = await ml_service.get_model_metrics(
                model_id=model.id,
                start_date=(datetime.now() - timedelta(days=7)).strftime("%Y-%m-%d"),
                end_date=datetime.now().strftime("%Y-%m-%d"),
            )
            logger.info(f"📈 Available metrics: {list(metrics.keys())}")

            # Example 9: Trigger model retraining
            logger.info("\n🔄 Triggering model retraining...")
            retrain_result = await ml_service.retrain_model(
                model_id=model.id,
                feedback_threshold=100,  # Retrain when 100 feedback samples available
            )
            logger.info(f"🔄 Retraining initiated: {retrain_result}")

            # Example 10: Get dashboard data
            logger.info("\n📊 Getting dashboard data...")
            dashboard = await ml_service.get_dashboard_data()
            logger.info(f"📋 Dashboard sections: {list(dashboard.keys())}")

        except Exception as e:
            logger.error(f"❌ Error: {e}")
            raise


if __name__ == "__main__":
    asyncio.run(main())
